### PR TITLE
Develop/podman module

### PR DESCRIPTION
**Pull request by n-hass**
### Description
 
- Starts from the rewrite of the module from @bamhm182
- Tries to zip extraConfig into declared options if they can be merged - if not, favours extraConfig. This is basically lib.zipAttrs if either value is already a list, otherwise lib.recursiveUpdate
- Adds some more edge cases to tests
- On available options, if a single value is acceptable but repeating values (ie port, network, environmentFile) are also ok, allow both single val and a list on the module option

### CODE REVIEW

File: modules/services/podman-linux/containers.nix [hunk 1]
```diff
@@ -223,55 +182,67 @@ in let
...
-          `ports = [ "8080:80" ];`
+          `exec = "sleep inf";`
         '';
       };
 
-      user = mkOption {
-        type = with types; nullOr (oneOf [ str int ]);
-        default = null;
-        description = "The user ID inside the container.";
+      extraOptions = mkOption {
```
Comment (n-hass) [inline modules/services/podman-linux/containers.nix]: @bamhm182 I’m not sure about this rename from extraPodmanArgs to extraOptions. in NixOS `virtualisation.oci-containers` it is extraOptions, but I’m undecided on if this is going too far on trying to be consistent. What do you think?

File: modules/services/podman-linux/podman-lib.nix [hunk 2]
```diff
@@ -1,71 +1,145 @@
...
+    listsAsDuplicateKeys = true;
+    mkKeyValue = normalizeKeyValue;
   };
-  serviceConfigType = with types; attrsOf (either primitive (listOf primitive));
 
-  unitConfigTypeRules = { After = with types; nullOr (listOf str); };
-  unitConfigDefaults = { After = null; };
-  unitConfigType = with types; attrsOf (either primitive (listOf primitive));
+  # meant for ini. favours b when two values are unmergeable
+  deepMerge = a: b:
```
Comment (n-hass) [inline modules/services/podman-linux/podman-lib.nix:28]: this should be achievable in equivalent but less code with a `lib.zipAttrsWith (impl)`. Do you think its worth writing it as that @bamhm182?

File: modules/services/podman-linux/containers.nix (hunk 1 above)
Comment (bamhm182) [reply modules/services/podman-linux/containers.nix]: I was going for extraPodmanArgs because the option is called "PodmanArgs", whereas `virtualisation.oci-containers` tries to be agnostic between docker podman and docker. I was also concerned that extraConfig and extraOptions sound too similar.

File: modules/services/podman-linux/podman-lib.nix (hunk 2 above)
Comment (bamhm182) [reply modules/services/podman-linux/podman-lib.nix:28]: Unless I'm missing it, I don't think the code difference is that large. I think it's fine the way you have it.

File: modules/services/podman-linux/containers.nix (hunk 1 above)
Comment (bamhm182) [reply modules/services/podman-linux/containers.nix]: That said, I think it's bike shedding to think too hard on this. I'm fine with either as both make sense. Happy to leave it as extraOptions as that's how it is within NixOS oci-containers.

**Comment by bamhm182** [comment]
Most recent commit looks good to me. No longer seeing anything glaring and it seems to work as expected. I would say we call it good here and try to get the main PR accepted.

File: modules/services/podman-linux/podman-lib.nix (hunk 2 above)
Comment (n-hass) [reply modules/services/podman-linux/podman-lib.nix:28]: Yeah it just makes it harder to follow 😓  was playing around with it. putting the result here for posterity. It would probably look something like this:

```nix
  mergeAttrs = with builtins; name: values:
    let
      realValues = filter (v: v != null) values;
      allAreAttrs = all (v: typeOf v == "set") realValues;
      anyIsList = any isList realValues;
    in
      if length realValues == 0 then
        null
      else if allAreAttrs then
        zipAttrsWith mergeAttrs realValues
      else if length realValues == 1 then
        head realValues
      else if anyIsList then
        let
          mergedList = concatLists (map (v:
            if isList v then v else [v]
          ) realValues);
        in
          if length mergedList == 0 then null else mergedList
      else
        last realValues;

  deepMerge = a: b: zipAttrsWith mergeAttrs [a b];
```

**Comment by n-hass** [comment]
Cool, thanks @bamhm182. I’ll revert the renaming from `extraPodmanArgs` to `extraOptions` and then merge.
